pkg/api/service.go

```diff
@@ -55,15 +64,63 @@ func (s *Service) Close() {
 
 func (s *Service) BatchSubscribeEnvelopes(
 	req *message_api.BatchSubscribeEnvelopesRequest,
-	server message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
+	stream message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
 ) error {
-	return status.Errorf(codes.Unimplemented, "method BatchSubscribeEnvelopes not implemented")
+	log := s.log.Named("subscribe")
+
+	// Send a header (any header) to fix an issue with Tonic based GRPC clients.
+	// See: https://github.com/xmtp/libxmtp/pull/58
+	err := stream.SendHeader(metadata.Pairs("subscribed", "true"))
+	if err != nil {
+		return status.Errorf(codes.Internal, "could not send header: %v", err)
+	}
+
+	requests := req.GetRequests()
+	if len(requests) == 0 {
+		return status.Errorf(codes.InvalidArgument, "missing requests")
+	}
+
+	ch, err := s.subscribeWorker.addListeners(requests)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid subscription request: %v", err)
+	}
+
+	var streamLock sync.Mutex
+	for exit := false; !exit; {
+		select {
+		case envs, open := <-ch:
+			if open {
+				func() {
+					streamLock.Lock()
+					defer streamLock.Unlock()
+					err := stream.Send(&message_api.BatchSubscribeEnvelopesResponse{
+						Envelopes: envs,
+					})
+					if err != nil {
+						log.Error("sending envelope to subscribe", zap.Error(err))
+					}
+				}()
+			} else {
+				// TODO(rich) Recover from backpressure
+				log.Info("stream closed due to backpressure")
+				exit = true
+			}
+		case <-stream.Context().Done():
+			log.Debug("stream closed")
+			exit = true
```

Comment:
You could also do this with a [labeled loop](https://www.ardanlabs.com/blog/2013/11/label-breaks-in-go.html) and a break

Reply:
or just explicitly return nil such as:
```

To get rid of the exit variable and simplify the loop termination, you can use a for loop combined with a return or break statement inside the select block. This approach avoids the need for an explicit exit variable and results in a cleaner code structure. Here's how you can rewrite the code:

Refactored Code:
go
Copy code
for {
    select {
    case envs, open := <-ch:
        if open {
            func() {
                streamLock.Lock()
                defer streamLock.Unlock()
                err := stream.Send(&message_api.BatchSubscribeEnvelopesResponse{
                    Envelopes: envs,
                })
                if err != nil {
                    log.Error("sending envelope to subscribe", zap.Error(err))
                }
            }()
        } else {
            // TODO: Recover from backpressure
            log.Info("stream closed due to backpressure")
            return nil  // Terminate the function when the channel is closed
        }

    case <-stream.Context().Done():
        log.Debug("stream closed")
        return nil  // Exit the function when the stream is done

    case <-s.ctx.Done():
        log.Info("service closed")
        return nil  // Exit when the service context is done
    }
}```